src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -115,13 +109,11 @@ private async Task<Solution> UpdateProjectsAsync(Solution solution, ChecksumColl
                 olds.Object.ExceptWith(newChecksums);
                 news.Object.ExceptWith(oldChecksums);
 
-                return await UpdateProjectsAsync(solution, olds.Object, news.Object, cancellationToken).ConfigureAwait(false);
-            }
+                using var _1 = PooledDictionary<ProjectId, ProjectStateChecksums>.GetInstance(out var oldMap);
+                using var _2 = PooledDictionary<ProjectId, ProjectStateChecksums>.GetInstance(out var newMap);
 
-            private async Task<Solution> UpdateProjectsAsync(Solution solution, HashSet<Checksum> oldChecksums, HashSet<Checksum> newChecksums, CancellationToken cancellationToken)
-            {
-                var oldMap = await GetProjectMapAsync(solution, oldChecksums, cancellationToken).ConfigureAwait(false);
-                var newMap = await GetProjectMapAsync(_assetProvider, newChecksums, cancellationToken).ConfigureAwait(false);
+                await PopulateOldProjectMapAsync().ConfigureAwait(false);
+                await PopulateNewProjectMapAsync(this).ConfigureAwait(false);
```

Comment:
inlined these helper methods to now be local functions in this main function.  moved to pooled dictionaries as well to help with allocation overhead.